ethereum/contracts/BeefyClient.sol

```diff
@@ -0,0 +1,474 @@
+     * @param nextAuthoritySetRoot merkle root of all public keys in that validator set
+     */
+    struct MMRLeaf {
+        uint8 version;
+        uint32 parentNumber;
+        bytes32 parentHash;
+        uint64 nextAuthoritySetId;
+        uint32 nextAuthoritySetLen;
+        bytes32 nextAuthoritySetRoot;
+        bytes32 parachainHeadsRoot;
+    }
+
+    /* State */
+
+    bytes32 public latestMMRRoot;
+    uint64 public latestBeefyBlock;
+
+    struct ValidatorSet {
+        uint256 id;
+        bytes32 root;
+        uint256 length;
+    }
+
+    ValidatorSet public currentValidatorSet;
+    ValidatorSet public nextValidatorSet;
+
+    uint256 public nextID;
+    mapping(uint256 => ValidationData) public validationData;
+
+    /* Constants */
+
+    // Used for calculating minimum number of required signatures
+    uint256 public constant THRESHOLD_NUMERATOR = 3;
+    uint256 public constant THRESHOLD_DENOMINATOR = 250;
+
+    uint64 public constant BLOCK_WAIT_PERIOD = 3;
+
+    /**
+     * @notice Deploys the BeefyClient contract
+     */
+    constructor() {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        nextID = 0;
+    }
+
+    // Once-off post-construction call to set initial configuration.
+    function initialize(
+        uint64 _startingBeefyBlock,
+        ValidatorSet calldata _initialValidatorSet,
+        ValidatorSet calldata _nextValidatorSet
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        latestBeefyBlock = _startingBeefyBlock;
+        currentValidatorSet = _initialValidatorSet;
+        nextValidatorSet = _nextValidatorSet;
+
+        // drop admin privileges
+        renounceRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    /* Public Functions */
+
+    /**
+     * @notice Executed by the incoming channel in order to verify commitment
+     * @param leafHash contains the merkle leaf to be verified
+     * @param proof contains simplified mmr proof
+     */
+    function verifyMMRLeafProof(bytes32 leafHash, MMRProof calldata proof)
+        external
+        view
+        returns (bool)
+    {
+        return MMRProofVerification.verifyLeafProof(latestMMRRoot, leafHash, proof);
+    }
+
+    /**
+     * @notice Executed by the prover in order to begin the process of block
+     * acceptance by the light client
+     * @param commitmentHash contains the commitmentHash signed by the validator(s)
+     * @param validatorClaimsBitfield a bitfield containing a membership status of each
+     * validator who has claimed to have signed the commitmentHash
+     * @param validatorSignature the signature of one validator
+     * @param validatorPosition the position of the validator, index starting at 0
+     * @param validatorPublicKey the public key of the validator
+     * @param validatorPublicKeyMerkleProof proof required for validation of the public key in the validator merkle tree
+     */
+    function newSignatureCommitment(
+        bytes32 commitmentHash,
+        uint64 validatorSetID,
```

Comment:
could this be just a boolean, like bool next true or false?

Reply:
obsolete.

Reply:
is it? still looks like it could be a boolean rather?

Reply:
its safer to use an ID instead of bool, so the verification code can discern between the following error cases and report properly:
1) The specified next validator set is no longer the currently-recorded next validator set.
2) The validator proof is incorrect.

Reply:
solidity code should optimize for gas, not for error reporting :P